### ISSUE TITLE
Builtins for `Map`

### ISSUE THREAD

**Issue by f-f** [dhall-lang/dhall-lang#979]
Usecase/context: we're designing a [Package Registry for PureScript](https://github.com/purescript/registry) in Dhall, and in the current iteration we're storing things like "dependencies" and "build targets" in a Dhall `Map`, because they are key-value mappings and we can't fix the keys beforehand so it's not possible to use a record.
However, we would need to "override" values inside the `Map` sometimes (see [this issue](https://github.com/purescript/registry/issues/17) for details), and this is not possible at all in Dhall right now.

So I was wondering if we'd be open to add some kind of manipulation builtin for `Map`? In our case having some kind of `Map/update` would fix the current concerns.
I think a good candidate for a builtin could be:
```dhall
Map/alter : 
  ∀(a : Type) → 
  ∀(k : Type) → 
  (Optional a → Optional a) → k → Map k a → Map k a  
```
..modeled after [`Data.Map.alter`](https://www.stackage.org/haddock/lts-15.10/containers-0.6.2.1/Data-Map-Lazy.html#v:alter), on top of which we could implement `insert`/`update`/`delete` in the Prelude.

If we had string comparison we could implement this with `List` operations, but adding such a builtin might be valuable anyways for performance reasons.
Of course if we add this then it could be exploited to implement string comparison, but IMO it's a fairly acceptable risk.

Thoughts?

**Comment by Gabriel439** [comment]
@f-f: I commented on the original issue with suggestions specific to your problem: https://github.com/purescript/registry/issues/17

... but I'll also provide a more general answer here.

In my view, the main issue with introducing `Map` operations is the same issue with introducing `Text` equality: they are both vulnerable to mistakes that are only detected at runtime (e.g. misspelled or missing keys).

It's also not just about the type safety: Dhall benefits a lot, ergonomically, from offloading almost all error-checking to the type system instead of doing error-checking at the term-level.  For example, it's the reason we've been able to go without `do` notation for `Optional` or an `Either`-like type for so long.

From the description on the original issue it sounds like record operations would be safer and more ergonomic, so I'd like to lean on making the record-based solution work.

**Comment by Nadrieril** [comment]
Also this would give us a way to test any two values for equality. The idea is that since `Map/alter` only changes the entry at a given key, if we take a map with only one key `x` and alter it with key `y`, depending on whether `x` and `y` are equal the map will be updated differently. We can use that to compute the equality of `x` and `y`.
Thus allowing this is equivalent to allowing an equality predicate, in which case we could have defined `alter` in terms of this equality anyways.

**Comment by ari-becker** [comment]
The description of the original issue reminds me of a related concern with Dhall at the moment, which is the question of modeling plugin archtiectures. I've run into this issue before with both Concourse (with its Resources) and with Terraform.

There are two naive ways of doing so - representing all of the plugin options in a giant union, or normalizing the plugin's options with `Prelude.JSON` (or to a `Map`, as the original issue presumes) when it comes time to render. I tried (in Concourse's case) the giant union option (adding additional plugins to the union as needed) and eventually ran into memory exhaustion issues, particularly when many of the plugin options wouldn't be used.

Currently we're also staring at the problem for our internal usecases - representing the individual configurations of individual microservices as "plugins" of a sort, so that we'll be able to add new microservices and their configurations as we go without needing to migrate types - and I've been procrastinating on it a bit because the support isn't really there.

If what I've described isn't clear enough, let me know and I'll take some time to draw up some examples.

**Comment by Gabriel439** [comment]
@f-f: What if there were a `fromMap` utility that let you convert a `Map` to the corresponding record literal if the value of the `Map` is known (i.e. "immediate")?  This would be similar to how `toMap` behaves but in reverse

Using the example from https://github.com/purescript/registry/issues/17, you might be able to write:

```dhall
somePackage
  with targets.lib.dependencies =
    toMap (fromMap somePackage.targets.lib.dependencies with effect = "^3.0.0")
```

It wouldn't be quite as ergonomic as:

```dhall
somePackage with targets.lib.dependencies.effect = "^3.0.0"
```

... but it seems like it would be close enough and also useful in other regards (e.g. trying to use Dhall to add structure to weakly-typed data)

**Comment by f-f** [comment]
@Gabriel439 oh yeah, restricting this to immediate values is clever, and it would be consistent with how we treat dependent equality right now.

Thought experiment: how would you feel about pushing it a little bit more on the sugar side?

I.e. why not making this syntax just work with `Map`s if they are immediate?

```dhall
somePackage with targets.lib.dependencies.effect = "^3.0.0"
```

I'm aware this means we'd need `//`, and the dot notation, and all the other record things to work as well, so I guess here I'm proposing "treating `Map` as a record if they are immediate", which makes it much nicer to use while keeping type safety high, and maybe making it even easier to standardize?

This would allow us to have "typechecked homogeneous records", which would be quite nice for other things as well.
E.g. another constraint we have for the Registry is something like "the `lib` target needs to have `sources = ["src/**/*.purs"]`". If we introduce this then we could just write a nice assertion in pure Dhall for that, something like

```dhall
let _ = somePackage.targets.lib.sources === [ "src/**/*.purs" ]
```

(where `targets : Map Text Target`)

**Comment by f-f** [comment]
This would get rid of some more awkwardness around our representation too.

E.g. right now we represent [a package set](https://github.com/purescript/registry/blob/master/v1/sets/20200418.dhall) as a record because we want to be able to override values, along the lines of `set // { effect = "3.0.0" }`.

But then we have to convert it to a `Map` because we need to be able to write down its type, e.g. [here](https://github.com/purescript/registry/blob/master/v1/Index.dhall):
```
let Map = (./Prelude.dhall).Map.Type

let Address = ./Address.dhall

in
  < Registry :  Map Text Address
  | PackageSet : 
    { compiler : Text
    , packages : Map Text Address
    }
  >
```

The awkwardness here is that users pull a package set and have to unpack/repack when they need to override things, as you can see in [this example](https://github.com/purescript/registry#overriding-a-package):

```dhall
let Registry = https://raw.githubusercontent.com/purescript/registry/master/v1/Registry.dhall

let upstream = https://raw.githubusercontent.com/purescript/registry/master/v1/sets/20200418.dhall

let overrides = { effect = From.Local ( ../my-effect/spago.dhall as Location ) }

in  Registry.Package::{
    , name = "my-package"
    , packages = 
        Registry.Index.PackageSet
          { compiler = upstream.compiler
          , packages = toMap (upstream.packages // overrides)    -- <- here
          }
    , targets =
        toMap
          { src = Registry.Target::{
            , sources = [ "src/**/*.purs" ]
            , dependencies = toMap { effect = "^2.0.0" }
            }
          }
    }
```

...but we could avoid this whole dance if one could somehow just override an immediate `Map`

**Comment by Gabriel439** [comment]
@f-f: Yeah, I like the idea of overloading `with` syntax to handle that.  The only minor change I would suggest is to make it explicit when that is happening, like this:

```dhall
somePackage with targets.lib.dependencies.(map).effect = "^3.0.0"
```

... or something similar.  That way it's much easier for us to specify the desugaring logic because we don't need to do type inference to detect if we need to apply the conversion.  The desugaring logic for `with` would only require a new judgement like:

```dhall
r with (map).ks… = toMap (fromMap r with ks)
```

I also thing it might be worth splitting that into two separate changes (one for `fromMap` and a separate proposal for `with` support).

**Comment by Nadrieril** [comment]
What does this mean for the value of the Map to be "immediate" ? Is that something straightforward to check during typechecking ?

**Comment by Gabriel439** [comment]
@Nadrieril: It's basically the same restriction as `toMap`, meaning that the argument cannot be abstract (e.g. a bound variable)

**Comment by Nadrieril** [comment]
@Gabriel439 I don't think I understand what you're saying. Typechecking only checks that the argument to toMap is a record. `\(x: { ... }) -> toMap x` is perfectly valid, isn't it ? It just won't normalize further unless applied to a record value. Here the problem is rather different: not all values of type `Map Text x` are valid inputs to fromMap. That's proper dependent typechecking. We will need to restrict the argument to be something that we can check has the right keys. But that's quite difficult, how do we do that without ruling out all interesting programs ?

**Comment by Gabriel439** [comment]
@Nadrieril: Yeah, you're right.  My mistake

I guess what I meant to say is that `fromMap` would require a record literal as an argument.  I still think that even with this restriction `fromMap` would be useful.

**Comment by f-f** [comment]
I assume that by "require a record literal" here we mean that it has the same semantics of [assertions](https://github.com/dhall-lang/dhall-lang/blob/master/standard/type-inference.md#assertions), i.e. it would work even if it's a variable, as long as it's "immediate".

So to recap, even with this restriction the following would work (note both `targets` and `dependencies` are `Map`s):
```dhall
somePackage with targets.(map).lib.dependencies.(map).effect = "^3.0.0"
```

How about my other example from above, which doesn't use `with` but only the dot notation?
Rewritten with the proposed syntax:
```dhall
let _ = assert : somePackage.targets.(map).lib.sources === [ "src/**/*.purs" ]
```

**Comment by Nadrieril** [comment]
Ok, if I understand correctly we would, at typechecking time, evaluate the argument of fromMap, and verify that it is a list literal containing records of the appropriate type, and whose mapKey fields are known. Otherwise it's a type error. That would be enough info to infer the type of the output of fromMap, so that a type annotation can most of the time be omitted. This in turn allows all sorts of magical back-and-forth like the `with` sugar you mention. Neat !

**Comment by ari-becker** [comment]
Proposal: what if, instead of:
```dhall
fromMap [ { mapKey = "foo" , mapValue = "bar" } ] === { foo = "bar" }
```
instead have something like the following?
```dhall
fromMap [ { mapKey = "foo" , mapValue = "bar" } ] 
  === { Type = { foo : Text }
      , parsed = { foo = "bar" }
      }
```
The idea being to allow the `Type` to be extracted so that it can be passed along in function parameters and other areas where the type must be provided for the record to be useful.

**Comment by Gabriel439** [comment]
@f-f: Yeah, we could also standardize that `(map)` can be used as a field accessor.

In fact, we could actually go a step further and use `.(map)` (or a different syntax, since that conflicts with field projection by type) *instead* of `fromMap` to convert something to a record.  It would actually make more sense to use a different name if we did that, though, like `.{record}`, to denote that we're obtaining a record as the result, not a `Map`

@ari-becker: I think if we provide a way to extract the type it should probably be provided as a separate utility rather than bundled within the same keyword

**Comment by Gabriel439** [comment]
Oh, whoops, I forgot that `.{record}` is valid syntax for projection by a set of fields.  But I think you all get the idea.  We just need some new field-access-like syntax for accessing the equivalent record representation.

**Comment by f-f** [comment]
How about something like `.^`?
The mnemonic would be "the arrow up is a 'cast'/'promotion' to a record".
Probably the mirrored version `^.` would be nice too, mnemonic: "cast up and access field"

Examples:

```dhall
somePackage with targets.^lib.dependencies.^effect = "^3.0.0"

somePackage with targets^.lib.dependencies^.effect = "^3.0.0"
```

**Comment by Gabriel439** [comment]
@f-f: Perhaps a minor variation on your suggestion, which is to have `.` on both sides of the `^`, like this:

```dhall
somePackage with targets.^.lib.dependencies.^.effect = "^3.0.0"
```

* so that the record itself can be accessed, too, like this:

  ```dhall
  somePackage with targets.^
  ```

* for consistency with other potential "components" (like `?` for `Optional` and `*` for `List`)

To expand on the latter point, we might eventually add `with` support for updating `List` or `Optional` fields (https://github.com/dhall-lang/dhall-lang/issues/924), like this:

```dhall
someRecord with foo.?.bar.*.baz = 1
```

... where `?` could be used to descend into an `Optional` value and `*` could be used to descend into a `List`.

I'm also considering another variation to use `/` as the punctuation for converting from a `Map` to a record (e.g. `somePackage with targets./.lib.dependencies./.effect`).  The advantages of using `/` are:

* If we ever want to convert from a record back to a `Map` we can use `\`
* The `/` symbol is closer to the baseline where the `.`s are so there is less of a visual disconnect

**Comment by Profpatsch** [comment]
Can we make this general enough to not be only useful in some corner cases? Stuff like

> we would, at typechecking time, evaluate the argument of fromMap, and verify that it is a list literal containing records of the appropriate type, and whose mapKey fields are known. Otherwise it's a type error.

gives me the jeebies. What about floating stuff out in `let`? What about hiding it behind an import? Would that still work? And can you factor stuff out into a function still?

I’m skeptical we can make this work without proper row types.

**Comment by Gabriel439** [comment]
@Profpatsch: Floating stuff out in a `let` would still work or hiding behind an import would still work.  This is essentially the same mechanism that powers how union constructors work (i.e. the `UnionType.Constructor` idiom), which requires statically knowing the union constructor at type-checking time.

**Comment by dominik-steenken** [comment]
It's been a while since this issue has been updated. I would find the feature discussed here *extremely* useful for my application. Is there any chance that this will be implemented soon?

**Comment by f-f** [comment]
I was thinking that maybe this could be something suitable for having a bounty on it?

**Comment by Gabriel439** [comment]
Yeah, I can put up a funding proposal for this as soon and mentor a contributor on this as soon as I'm done mentoring GSoC

**Comment by Gabriel439** [comment]
Alright, I put up a funding proposal for a related change (for `Optional`, instead of `Map`) here: https://discourse.dhall-lang.org/t/expense-proposal-add-with-support-for-modifying-optional-values/520

My reasoning is that this will help get the ball rolling and I think standardizing support for `with` working on `Optional`s is higher priority at the moment, especially for Kubernetes support

**Comment by dominik-steenken** [comment]
Thank you! Hopefully this will revive this issue somewhat. If i ever get to a less hectic place in my current project, i might help out myself.

**Comment by f-f** [comment]
Update on this: #1254 implements support for using the `with` syntax with `Optional` through the `?` symbol, which as noted above will help get the ball rolling on this
